x-pack/plugins/security_solution/public/detections/containers/detection_engine/rules/use_rule_from_timeline.tsx

```diff
@@ -0,0 +1,223 @@
+} from '@kbn/timelines-plugin/public';
+import { useDispatch } from 'react-redux';
+
+import { i18n } from '@kbn/i18n';
+import { useAppToasts } from '../../../../common/hooks/use_app_toasts';
+import { useSourcererDataView } from '../../../../common/containers/sourcerer';
+import type { TimelineModel } from '../../../..';
+import type { FieldValueQueryBar } from '../../../components/rules/query_bar';
+import { sourcererActions } from '../../../../common/store/sourcerer';
+import {
+  dispatchUpdateTimeline,
+  queryTimelineById,
+} from '../../../../timelines/components/open_timeline/helpers';
+import { useGetInitialUrlParamValue } from '../../../../common/utils/global_query_string/helpers';
+import { buildGlobalQuery } from '../../../../timelines/components/timeline/helpers';
+import { getDataProviderFilter } from '../../../../timelines/components/timeline/query_bar';
+import { SourcererScopeName } from '../../../../common/store/sourcerer/model';
+
+export const RULE_FROM_TIMELINE_URL_PARAM = 'createRuleFromTimeline';
+
+export interface RuleFromTimeline {
+  loading: boolean;
+  onOpenTimeline: (timeline: TimelineModel) => void;
+}
+
+export const initialState = {
+  index: [],
+  queryBar: {
+    query: { query: '', language: 'kuery' },
+    filters: [],
+    saved_id: null,
+  },
+};
+
+type SetRuleQuery = ({
+  index,
+  queryBar,
+}: {
+  index: string[];
+  queryBar: FieldValueQueryBar;
+}) => void;
+
+/**
+ * When returned property updated === true,
+ * the index and queryBar properties have been updated from timeline data
+ * queried either from id in the url param or by passing a timeline to returned callback onOpenTimeline
+ */
+export const useRuleFromTimeline = (setRuleQuery: SetRuleQuery): RuleFromTimeline => {
+  const dispatch = useDispatch();
+  const { addError } = useAppToasts();
+
+  // selectedTimeline = timeline to set rule from
+  const [selectedTimeline, setRuleFromTimeline] = useState<TimelineModel | null>(null);
+
+  // start RuleFromTimeline definition
+  const [loading, setLoading] = useState(true);
+
+  const onOpenTimeline = useCallback(
+    (timeline: TimelineModel) => {
+      setRuleFromTimeline(timeline);
+
+      if (!isEmpty(timeline.indexNames)) {
+        // let sourcerer manage the selected browser fields by setting timeline scope to the selected timeline data view
+        // sourcerer handles the logic of if the fields have been fetched or need to be fetched
+        dispatch(
+          sourcererActions.setSelectedDataView({
+            id: SourcererScopeName.timeline,
+            selectedDataViewId: timeline.dataViewId,
+            selectedPatterns: timeline.indexNames,
+          })
+        );
+      }
+    },
+    [dispatch]
+  );
+  // end RuleFromTimeline definition
+
+  // start browser field management
+  const { browserFields, dataViewId, selectedPatterns } = useSourcererDataView(
+    SourcererScopeName.timeline
+  );
+
+  const [ogDataView] = useState({ dataViewId, selectedPatterns });
```

Comment:
why is this in useState if the update function is unused?

Reply:
also consider renaming to just originalDataView or something

Reply:
this way when dataViewId and selectedPatterns update, `ogDataView` will stay the original values. Do you know another way to do this?

Reply:
good use case for `useRef`?

Reply:
renamed it tho